google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+                      return this->RetryLoop(retry_policy, backoff_policy,
+                                             retry_type);
+                    }
+                    std::lock_guard<std::mutex> g{this->mu_};
+                    this->impl_ = std::move(*reinitialize_response);
+                    this->current_status_ = Status(
+                        StatusCode.kDataLoss, "Stream failed, May try again");
+                    if (retry_type == RetryType.Read) {
+                      return make_ready_future(absl::optional());
+                    }
+                    if (retry_type == RetryType.Status) {
+                      return make_ready_future(true);
+                    }
+                    return make_ready_future(false);
+                  });
+            });
+          });
+    }
+    current_status_ = Status(StatusCode.kUnavailable, "Permanent Error");
+    if (retry_type == RetryType.Read) {
+      return make_ready_future(absl::optional());
+    }
+    if (retry_type == RetryType.Status) {
+      return make_ready_future(true);
+    }
+    return make_ready_future(false);
+  }
+
+  void ProcessRead() {
+    read_future_ = make_optional(impl_->Read().then(
+        [this](future<absl::optional<Response>> optional_response_future) {
+          this->read_future_ = absl::optional();
+          absl::optional<Response> optional_response =
+              optional_response_future.get();
+          if (optional_response.has_value()) {
+            return make_ready_future(optional_response);
+          }
+          std::shared_ptr<RetryPolicy const> retry_policy =
+              this->retry_policy_prototype_->clone();
+          std::shared_ptr<BackoffPolicy const> backoff_policy backoff_policy =
+              this->backoff_policy_prototype_->clone();
+          return this->RetryLoop(retry_policy, backoff_policy, RetryType.Read);
+        }));
+  }
+
+  void ProcessWrite(Request const& r, grpc::WriteOptions o) {
+    write_future_ =
+        make_optional(impl_->Write(r, o).then([this, r](future<bool> write_fu) {
+          this->write_future_ = absl::optional();
+          if (write_fu.get()) {
+            this->write_future_ = absl::optional();
+            return make_ready_future(true);
+          }
+          std::shared_ptr<RetryPolicy const> retry_policy =
+              this->retry_policy_prototype_->clone();
+          auto const backoff_policy = this->backoff_policy_prototype_->clone();
+          return this->RetryLoop(retry_policy, backoff_policy, RetryType.Write);
+        }));
+  }
+
+  std::unique_ptr<RetryPolicy const> const retry_policy_prototype_;
+  std::unique_ptr<BackoffPolicy const> const backoff_policy_prototype_;
+  AsyncSleeper sleeper_;
+  StreamFactory<ResponseType, RequestType> const stream_factory_;
+  StreamReinitializer<ResponseType, RequestType> const reinitializer_;
+  std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>> impl_;
```

Comment:
impl needs to be guarded by mu_- can you 1) put all variables guarded by the mutex below the mutex and separated by a newline and 2) put a comment // ABSL_GUARDED_BY(mu) after the semicolon for all of these variables which are mutex guarded as an annotation?